src/contract.rs

```diff
@@ -0,0 +1,613 @@
+use crate::{ft_messages::*, utils::*};
+use dao_io::*;
+use gstd::{exec, msg, prelude::*, ActorId, String};
+
+pub const BASE_PERCENT: u8 = 100;
+
+static mut DAO: Option<Dao> = None;
+
+#[async_trait::async_trait]
+pub trait DaoHandler {
+    fn add_to_whitelist(&mut self, member: &ActorId);
+
+    async fn submit_membership_proposal(
+        &mut self,
+        transaction_id: Option<u64>,
+        applicant: &ActorId,
+        token_tribute: u128,
+        shares_requested: u128,
+        quorum: u128,
+        details: String,
+    );
+
+    fn submit_funding_proposal(
+        &mut self,
+        applicant: &ActorId,
+        amount: u128,
+        quorum: u128,
+        details: String,
+    );
+
+    fn submit_vote(&mut self, proposal_id: u128, vote: Vote);
+
+    async fn process_proposal(&mut self, transaction_id: Option<u64>, proposal_id: u128);
+
+    async fn ragequit(&mut self, transaction_id: Option<u64>, amount: u128);
+
+    async fn abort(&mut self, transaction_id: Option<u64>, proposal_id: u128);
+
+    fn set_admin(&mut self, new_admin: &ActorId);
+
+    fn update_delegate_key(&mut self, new_delegate_key: &ActorId);
+
+    async fn continue_transaction(&mut self, transaction_id: u64);
+}
```

Comment:
This hack is quite excessive. Try to move the `Dao` struct to this file, make it private, and make these methods for it.

Reply:
> This hack is quite excessive. Try to move the `Dao` struct from `io/src/lib.rs` to this file, make it private, and make these methods for it.

Probably in terms of resources this variant is not effective, but good educational example. Maybe make sense in app-state separation of concerns.

Reply:
I don't really understand what educational example you mean. You already separated public and private states here:
https://github.com/gear-dapps/dao/blob/4650d0ee676a7f7de0d2ce34698946ee7ff8950e/io/src/lib.rs#L18-L56

But the first one is useless in the `io` crate because it contains the `HashMap` type that can't be returned from a contract due restrictions of the SCALE codec. So it should be moved to `src/lib.rs` and be private. After that, this trait won't be needed.